### PR TITLE
Add check that string read is not more than max length

### PR DESCRIPTION
**Pull request by amgross**
# Pull request

**Choose Correct**

- [X] bug
- [ ] feature

**Describe the pull request**

Add fix to #326 , when passing string with max length set, ensure the real length not exceeds max length 
**To Reproduce**

Run erpcgen on:
[example.erpc.txt](https://github.com/EmbeddedRPC/erpc/files/10111270/example.erpc.txt)

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**

- OS: Windows
- eRPC Version:1.9.1

**Steps you didn't forgot to do**

- [X] I checked if other PR isn't solving this issue.
- [X] I read Contribution details and did appropriate actions.
- [X] PR code is tested.
- [X] PR code is formatted.
- [X] Allow edits from maintainers pull request option is set (recommended).

**Additional context**

I made just partially fix, need help for continue. What missing:
1. `{%  if info.stringAllocSize != info.stringLocalName + "_len" %}` is not working, don't know how to make it working
2. Don't know how to fix client side

### CODE REVIEW

**Comment by amgross** [comment]
@Hadatko as you can see, the compilation failed due to point 1 I mentioned in **Additional context**, can you help me and fix my commit? I don't know how to do it

**Comment by amgross** [comment]
BTW, the fix is based on the decodeBinaryType template

**Comment by Hadatko** [comment]
Ok thank you. I will take a look into this.

**Comment by amgross** [comment]
I added fix to client side, and added the checks also in the read/write binary (was exist just in server write binary).
In my view need just to see how to fix the `{% if info.stringAllocSize != info.stringLocalName + "_len" %}`

**Comment by amgross** [comment]
Hi @Hadatko ,
The fix is fully ready for code review.
The failures in the tests are: `FAILED to find pattern 'codec\->writeString\(strlen\(\(const\s*char\*\)data\->a\),\s*\(const\s*char\*\)data\->a\)`
Because I changed the `strlen` to be outside the `writeString` so I can reuse the return value to compare it against the max value.

**Comment by Hadatko** [comment]
Can you fix the test too? Remove line, add line.

**Comment by amgross** [comment]
I will try, to where the failure page gone?

**Comment by amgross** [comment]
@Hadatko can you point me in which tests the YML failed?

**Comment by amgross** [comment]
@Hadatko I done (hopefully)

**Comment by amgross** [comment]
@Hadatko @MichalPrincNXP can you prioritize this PR to version 1.10.0?
It suppose to prevent buffer overflow that the client may cause on the server (and maybe also vise versa)

**Comment by Hadatko** [comment]
@amgross I put note here. Likely we do not manage to get it into 1.10.0 but into 1.11.0 it shouldn't be an issue. On develop branch it will be present once @MichalPrincNXP will have time, but we need to make decision on note i added.

**Comment by amgross** [comment]
> but we need to make decision on note i added

@Hadatko what is the note you are referring to?

**Comment by Hadatko** [comment]
image

**Comment by amgross** [comment]
@Hadatko You probably didn't submitted your review, see at: [github documentation](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/reviewing-proposed-changes-in-a-pull-request) that "Before you submit your review, your line comments are pending and only visible to you".
image

**Comment by Hadatko** [comment]
From my point of view asserts are good to replace conditions in release target (if you would write if condition it is present in DEBUG and in RELEASE target. Using assert is only in DEBUG target and halts system). 
Reason to set assert for `write` cases is that you as an owner are trying to writting value which can be out of range.
Reason to use updateStatus+condition for `read` cases is that you can receive wrong data from communication channel.

File: erpcgen/src/templates/c_coders.template [hunk 1]
```diff
@@ -228,7 +244,23 @@ codec->readData({$info.name}, {$info.sizeTemp} * sizeof({$info.builtinTypeName})
...
-codec->writeString(strlen((const char*){$info.name}), (const char*){$info.name});
+{
+    uint32_t {$info.stringLocalName}_len = strlen((const char*){$info.name});
+{% if info.stringAllocSize != info.stringLocalName & "_len"  %}
+
+{% if source == "server" %}
+    erpc_assert({$info.stringLocalName}_len <= {$info.stringAllocSize});
+
+{%  else %}
+    if ({$info.stringAllocSize} <  {$info.stringLocalName}_len)
```
[the remark this answers is not in the document]
Comment (amgross) [reply erpcgen/src/templates/c_coders.template:255]: OK, got your point. I agree and I will fix it.

[the remark this answers is not in the document]
Comment (amgross) [reply erpcgen/src/templates/c_coders.template]: @Hadatko I fixed it, can you please re-review?

File: erpcgen/src/templates/c_coders.template [hunk 2]
```diff
@@ -239,8 +263,9 @@ codec->write({$info.name});
 {% enddef --------------------------------------- BuiltinType %}
 
 {% def encodeBinaryType(info) -----------------  %}
-{% if source == "server" && info.size != info.maxSize %}
-erpc_assert({$info.size} <= {$info.maxSize} * sizeof({$info.mallocSizeType}));
+{% if info.sizeTemp != info.maxSize %}
+erpc_assert({% if source == "client" && info.pointerScalarTypes %}*{% endif %}{$info.size} <= {$info.maxSize} * sizeof({$info.mallocSizeType}));
+
```
Comment (Hadatko) [inline erpcgen/src/templates/c_coders.template:268]: one empty newline can be removed

File: erpcgen/src/templates/c_coders.template [hunk 3]
```diff
@@ -228,7 +244,15 @@ codec->readData({$info.name}, {$info.sizeTemp} * sizeof({$info.builtinTypeName})
 {# Encode sending data #}
 {% def encodeBuiltinType(info) -----------------  %}
 {% if info.builtinType == "kStringType" %}
-codec->writeString(strlen((const char*){$info.name}), (const char*){$info.name});
+{
+    uint32_t {$info.stringLocalName}_len = strlen((const char*){$info.name});
+{% if info.stringAllocSize != info.stringLocalName & "_len"  %}
+
+    erpc_assert({$info.stringLocalName}_len <= {$info.stringAllocSize});
+
```
Comment (Hadatko) [inline erpcgen/src/templates/c_coders.template:252]: move one empty line outside of condition

**Review by Hadatko — CHANGES_REQUESTED** [review]

**Comment by Hadatko** [comment]
@amgross Just few minor changes and it is ready for merge.

File: erpcgen/src/templates/c_coders.template (hunk 2 above)
Comment (amgross) [reply erpcgen/src/templates/c_coders.template:268]: I thought it is better with it, but whatever, I will do it here and in the write string

File: erpcgen/src/templates/c_coders.template (hunk 3 above)
Comment (amgross) [reply erpcgen/src/templates/c_coders.template:252]: agree

File: erpcgen/src/templates/c_coders.template [hunk 4]
```diff
@@ -228,7 +244,14 @@ codec->readData({$info.name}, {$info.sizeTemp} * sizeof({$info.builtinTypeName})
 {# Encode sending data #}
 {% def encodeBuiltinType(info) -----------------  %}
 {% if info.builtinType == "kStringType" %}
-codec->writeString(strlen((const char*){$info.name}), (const char*){$info.name});
+{
+    uint32_t {$info.stringLocalName}_len = strlen((const char*){$info.name});
+
+{% if info.stringAllocSize != info.stringLocalName & "_len"  %}
+    erpc_assert({$info.stringLocalName}_len <= {$info.stringAllocSize});
```
Comment (Hadatko) [inline erpcgen/src/templates/c_coders.template:251]: you moved one top and one removed :D
my point was to change
```template
{% if info.stringAllocSize != info.stringLocalName & "_len"  %}

    erpc_assert({$info.stringLocalName}_len <= {$info.stringAllocSize});

{% endif %}
```
to 

```template
{% if info.stringAllocSize != info.stringLocalName & "_len"  %}

    erpc_assert({$info.stringLocalName}_len <= {$info.stringAllocSize});
{% endif %}
//newline was moved here outside
```
or

```template
//newline was moved here outside
{% if info.stringAllocSize != info.stringLocalName & "_len"  %}
    erpc_assert({$info.stringLocalName}_len <= {$info.stringAllocSize});

{% endif %}
```

Comment (Hadatko) [reply erpcgen/src/templates/c_coders.template:251]: You need think how it looks when condition is true and when is false.

**Review by Hadatko — APPROVED** [review]

File: erpcgen/src/templates/c_coders.template (hunk 4 above)
Comment (amgross) [reply erpcgen/src/templates/c_coders.template:251]: I thought you will want no line between the assert to the codec->writeString. Such as you asked me to remove the line between the other assert to the codec->writeBinary. I am probably missing something, anyway I done it.

**Comment by amgross** [comment]
@Hadatko , I think I did now what you asked

**Comment by amgross** [comment]
@MichalPrincNXP any comments?
@Hadatko finished reviewing.

**Review by Hadatko — APPROVED** [review]

**Comment by Hadatko** [comment]
@amgross i think @MichalPrincNXP will be available in beginning of new year. This is not critical bug so it will be likely added to next release as current one is in feature freeze state.

**Comment by amgross** [comment]
OK, I will wait,
In my point of view it is critical issue because the bug here may lead to buffer overflow which one attacker side can cause the other, but I understand it is not yours point of view.

**Comment by Hadatko** [comment]
@amgross Thank you for your patience

**Review by MichalPrincNXP — APPROVED** [review]

**Comment by MichalPrincNXP** [comment]
Thank you, @amgross , @Hadatko  for the effort and the patience!

**Pull request merged**